### PR TITLE
feat(rust, python): The big error revamp

### PR DESCRIPTION
**Pull request by aldanor**
This turned to be a bit bigger than was initially planned 🙈 But I figured if I started it, should probably finish it.

Here's a few precursory points:
- Perceived maturity of any library is greatly affected by the quality of its error messages (at least subjectively so)
- In polars, error message consistency was a bit of a problem:
  - Some error messages were like 'idx out bounds', some were 10 lines long with detailed explanations
  - Some were formed like sentences, some were lowercase, some were not, some ended with trailing period, some didn't, etc
  - Data types were called datatype, dtypes, Dtypes, DataType, Datatypes, etc. Same with e.g. dataframes
  - Various things and names were single/double/back/non-quoted at random
- Another point is that, given the fallible nature of the library, there's no consistent and easy way for developers to work with errors (other than doing `format!().into()` everywhere and a ton of it statements)

Here's a list of proposed changes:
- Removed `anyhow` dependency everywhere (was already marked as "TODO: remove")
- Clean up error variants, their str reprs, etc
- Use `Cow<'static, str>` inside `ErrString` directly (also widens conversion range)
- Add `polars_err!()`, `polars_bail!()` `polars_ensure!()` macros and use them (somewhat similar to anyhow/eyre libs)
- `polars_ensure!()` allows to write conditions in positive form (i.e. "ensure that X" instead of having to flip it like "if not X")
- Using those macros shaved off more than 1k lines of pure boilerplate code
- All error messages and all error-related code (~600 occurrences) was refactored, error messages reviewed to try and bring them to be more consistent overall
- Potential perf effect: error macros yield `#[cold]` branches which may improve branching downstream (but definitely shouldn't worsen it)
- Some error messages were wrong because they were copied/pasted from a wrong place
- Cleaned up LogicalPlan formatting – no longer have preceding or trailing whitespace (especially when it's 1 line and is wrapped in an error message)
- Moved commonly copied/pasted error messages into macros (e.g. append/extend/oob etc)
- Removed a multitude of error-related macros, helpers and functions all over the place
- I think I've fixed a bug spotted in `impl_set_at_idx_with!()`, `set_at_idx` (was checking > instead of >=), someone should verify
- I think I've also fixed a bug spotted in range checks for literal expressions (e.g. negative bounds for u32s), someone should verify

### CODE REVIEW

**Comment by aldanor** [comment]
Side notes:
- This will obviously need to be rebased (over the last 10 commits or so). Every rebase will be a _major_ pain for this PR, so if the intention is to eventually merge it, it would have to be done reasonably quick
- One of the next steps at some point might be reviewing all uses of `panic!()` (and then, of unwraps/expects). I believe there's a multitude of cases where explicit panics are used in already-fallible functions where just returning an Err would suffice.

File: polars/polars-core/src/chunked_array/ops/set.rs [hunk 1]
```diff
@@ -8,22 +8,12 @@ use crate::utils::{align_chunks_binary, CustomIterTools};
 
 macro_rules! impl_set_at_idx_with {
     ($self:ident, $builder:ident, $idx:ident, $f:ident) => {{
-        let mut idx_iter = $idx.into_iter();
         let mut ca_iter = $self.into_iter().enumerate();
 
-        while let Some(current_idx) = idx_iter.next() {
-            if current_idx as usize > $self.len() {
```
Comment (aldanor) [inline polars/polars-core/src/chunked_array/ops/set.rs:15]: > or >=?

File: polars/polars-core/src/chunked_array/ops/set.rs [hunk 2]
```diff
@@ -210,19 +189,10 @@ impl<'a> ChunkSet<'a, &'a str, String> for Utf8Chunked {
         let mut ca_iter = self.into_iter().enumerate();
         let mut builder = Utf8ChunkedBuilder::new(self.name(), self.len(), self.get_values_size());
 
-        for current_idx in idx_iter {
-            if current_idx as usize > self.len() {
```
Comment (aldanor) [inline polars/polars-core/src/chunked_array/ops/set.rs:214]: `>` or `>=`?

File: polars/polars-lazy/src/physical_plan/expressions/literal.rs [hunk 3]
```diff
@@ -58,21 +62,18 @@ impl PhysicalExpr for LiteralExpr {
                     ca.into_inner().into_series()
                 }
                 DataType::UInt32 => {
-                    if *low >= 0 || *high <= u32::MAX as i64 {
```
Comment (aldanor) [inline polars/polars-lazy/src/physical_plan/expressions/literal.rs:61]: Looks like this was backwards

File: polars/polars-core/src/series/series_trait.rs [hunk 4]
```diff
@@ -149,39 +139,37 @@ pub(crate) mod private {
...
+            invalid_operation_panic!(zip_outer_join_column, self)
         }
 
         fn subtract(&self, _rhs: &Series) -> PolarsResult<Series> {
-            invalid_operation_panic!(self)
+            invalid_operation_panic!(sub, self)
         }
         fn add_to(&self, _rhs: &Series) -> PolarsResult<Series> {
-            invalid_operation!(self)
+            invalid_operation_panic!(add, self)
```
Comment (aldanor) [inline polars/polars-core/src/series/series_trait.rs:149]: Btw, why all the panics here when most of these methods return a Result?

File: polars/polars-core/src/chunked_array/ops/set.rs (hunk 1 above)
Comment (ritchie46) [reply polars/polars-core/src/chunked_array/ops/set.rs:15]: Yeap, that should be `>=` indeed. :+1:

File: polars/polars-lazy/src/physical_plan/expressions/literal.rs (hunk 3 above)
Comment (ritchie46) [reply polars/polars-lazy/src/physical_plan/expressions/literal.rs:61]: Yeap, shows the benefit of reasoning with `ensure`. :)

File: polars/polars-core/src/series/series_trait.rs (hunk 4 above)
Comment (ritchie46) [reply polars/polars-core/src/series/series_trait.rs:149]: Legacy reasons. If they return `Result` we should make use of that.

**Review by ritchie46 — COMMENTED** [review]
This is a great improvement. :raised_hands: 

I only think we should make sentences of our error messages. By that I mean start with a capital letter and end with a dot. This keeps it consistent with multi-line error messages that do more explaining.

**Comment by stinodego** [comment]
This is huge! What an effort.

> if the intention is to eventually merge it, it would have to be done reasonably quick

We could pick up any minor nitpicks in separate PRs so that we can avoid major conflicts.

> I only think we should make sentences of our error messages. By that I mean start with a capital letter and end with a dot. This keeps it consistent with multi-line error messages that do more explaining.

I agree with this one!

File: polars/polars-core/src/series/series_trait.rs (hunk 4 above)
Comment (aldanor) [reply polars/polars-core/src/series/series_trait.rs:149]: Yea, should be a separate thing – to eventually get rid of as many panics (and unwraps) as possible, the first thing being panics in fallible functions. There's not many explicit panics though (compared to the number of error checks), maybe just a few hundred :) 

// There's also an enormous amount of code like e.g. "check if not none; then unwrap", or "check some condition; then unwrap" – in many of those cases unwrap is not needed if code is phrased a little differently.

Comment (ritchie46) [reply polars/polars-core/src/series/series_trait.rs:149]: > // There's also an enormous amount of code like e.g. "check if not none; then unwrap", or "check some condition; then unwrap" – in many of those cases unwrap is not needed if code is phrased a little differently.

Yeap, the `if let Some() =` is great there. Some `unwrap()`s should stay IMO, there are good assertion and blow up in our face, so we can find the bugs.

**Comment by aldanor** [comment]
> I only think we should make sentences of our error messages. By that I mean start with a capital letter and end with a dot. This keeps it consistent with multi-line error messages that do more explaining.

I know I'm in the minority here so there's little chances, but let me just try to defend my point 🙂 

---

In Rust compiler itself, errors are always lowercase sentences. Explanations that follow below (unless they are inline annotations to the code, like lifetime tips), will be typically sentence-style (capital-cased and with a trailing period). E.g.:

```rust
error[E0070]: invalid left-hand side of assignment
  --> src/main.rs:13:7
   |
13 |     1 = 2;
   |     - ^
   |     |
   |     cannot assign to this expression

For more information about this error, try `rustc --explain E0070`.
```

Subjectively, this looks great, stylish and less noisy than ```error[E0070]: Invalid left-hand side of assignment.```

Maybe I'm just being too spoiled by the awesome rustc errors (hats off to compiler team) :)

---

CPython source itself is very inconsistent, depending on who has written the code:

```bash
cpython on main [?]
> rg -o 'raise [A-Z]\w+\([\'"]([a-z]).*' | wc -l
1615  # exception starts with lowercase
❯ rg -o 'raise [A-Z]\w+\([\'"]([A-Z]).*' | wc -l
935   # exception starts with uppercase
```

---

For Python exceptions, you will always see the type of the exception first, which will already be uppercase (most likely `ComputeError` in polars).

Same as with rustc compiler example above, I think there's nothing wrong with having the exception being lowercase since it reads just fine even if there are "sentence-style" explanations following below.

```python
>>> raise RuntimeError(
    "invalid left-hand side of assignment\n\nFor more information about this error, open a PR."
)
RuntimeError: invalid left-hand side of assignment

For more information about this error, open a PR.
```

I think the above looks quite nice and succinct already, somewhat rustc-style. I also agree that to keep it nice and consistent, the optional "explanations" below need to be separated by a newline and should be formatted "sentence-style", with trailing period and such.

Maybe a nicer way to do it would be to extend the error type to allow the 'chain of errors' (kind of like how eyre does it) instead of the manual `format!("{}\n\n{}\n\n{}")`, so that we have a consistent control over it, globally in the entire codebase.

---

Trailing newlines and dots. I think including any trailing newlines and periods in the error messages themselves may not be the best idea unless you're sure that's the final message. You can always do it as the very _last_ step if really needed (e.g. when converting PolarsError to PyPolarsError if you really want it), or don't do it at all.

In this PR, the logical plan formatting was also fixed, so that instead of this:
```
Error originated from this expression: '  DF [ ...] FOO; BAR
'
```
you get this
```
Error originated from this expression: 'DF [ ...] FOO; BAR'
```

So the choice would be really in doing it titlecase or not, where I would personally vote for a succinct rustc-style (first line lowercase; explanations below sentence-style).

**Comment by ritchie46** [comment]
> So the choice would be really in doing it titlecase or not, where I would personally vote for a succinct rustc-style (first line lowercase; explanations below sentence-style).

Ok, you have convinced me. rustc's error messages are the best I've experienced in any tool, so let's follow that example and try to setup a consistent architecture for this.

**Comment by aldanor** [comment]
Rebased on master, so should be good to go unless there's any particular comments.

// also wondering whether half a thousand of new `#[cold]` tags would help llvm at least a tiny bit with branching or not... 🙂

**Comment by ritchie46** [comment]
> // also wondering whether half a thousand of new #[cold] tags would help llvm at least a tiny bit with branching or not... 

Me too! :)

Can you do a rebase and ping me again? Something got between it.

**Comment by aldanor** [comment]
> Can you do a rebase and ping me again? Something got between it.

@ritchie46 Done

**Pull request merged**